src/main/scala/cromwell/engine/io/gcs/NioGcsPath.scala

```diff
@@ -0,0 +1,162 @@
+package cromwell.engine.io.gcs
+
+import java.io.File
+import java.net.URI
+import java.nio.file.WatchEvent.{Kind, Modifier}
+import java.nio.file._
+import java.util
+
+import scala.collection.JavaConverters._
+import scala.language.implicitConversions
+import scala.language.postfixOps
+import scala.util.Try
+
+object NioGcsPath {
+  def apply(path: String)(implicit gcsFileSystem: GcsFileSystem)= {
+    gcsFileSystem.getPath(path)
+  }
+
+  implicit class PathEnhanced(val path: Path) extends AnyVal {
+    def asGcsPath(implicit gcsFileSystem: GcsFileSystem) = path match {
+      case gcsPath: NioGcsPath => gcsPath
+      case otherPath: Path => gcsFileSystem.getPath(otherPath.toString).asInstanceOf[NioGcsPath]
+      case _ => throw new IllegalArgumentException("Only GcsPaths are supported.")
+    }
+  }
+
+  val protocol = GcsFileSystem.Protocol
+}
+
+/**
+  * NOTE: Currently called NioGcsPath so it can exist alongside the current GcsPath class.
+  *       If this approach was to be validated the current GcsPath class would be replaced by this one.
+  * This class proposes an implementation of the java.nio.Path interface for GoogleCloudStorage.
+  * The following methods are yet to be implemented:
+  *   relativize
+  *   compareTo
+  * @param chunks array containing all parts of the path in between separators - except the protocol (gs://)
+  *               eg: gs://path/to/resource.txt -> chunks = [path, to, resource.txt]
+  * @param absolute true if this path is to be considered absolute.
+  *                 Only absolute GCS paths can be used to actually locate resources.
+  *                 Calling methods on an absolute path can return a relative paths (eg subpath).
+  * @param gcsFileSystem the gcsFileSystem to be used when performing operations on this path
+  */
+class NioGcsPath(private val chunks: Array[String], absolute: Boolean)(implicit gcsFileSystem: GcsFileSystem) extends Path {
+  import NioGcsPath._
+
+  private val separator = GcsFileSystem.Separator
+
+  private val objectChunks = chunks match {
+    case values if isAbsolute && values.nonEmpty => values.tail
+    case _ => chunks
+  }
+
+  private val fullPath = chunksToString(chunks)
+
+  // Attributes
```

Comment:
I'd ask to either flesh this comment out or remove it. It doesn't mean anything to me as is